aws-datastore/src/test/java/com/amplifyframework/datastore/storage/InMemoryStorageAdapter.java

```diff
@@ -81,21 +81,36 @@ public void initialize(
     public <T extends Model> void save(
             @NonNull final T item,
             @NonNull final StorageItemChange.Initiator initiator,
-            @NonNull final QueryPredicate predicate,
+            @Nullable final QueryPredicate predicate,
             @NonNull final Consumer<StorageItemChange.Record> onSuccess,
             @NonNull final Consumer<DataStoreException> onError
     ) {
-        items.add(item);
-        StorageItemChange.Record save = StorageItemChange.<T>builder()
-                .item(item)
-                .itemClass((Class<T>) item.getClass())
-                .type(StorageItemChange.Type.SAVE)
-                .predicate(predicate)
-                .initiator(initiator)
-                .build()
-                .toRecord(storageItemChangeConverter);
-        changeRecordStream.onNext(save);
-        onSuccess.accept(save);
+        final int index = indexOf(item);
+        if (index > -1) {
+            // Update
+            Model savedItem = items.get(index);
+            if (predicate == null || predicate.evaluate(savedItem)) {
+                items.remove(index);
+                save(item, initiator, predicate, onSuccess, onError);
+                return;
+            }
+            onError.accept(new DataStoreException(
+                    "Conditional check failed.",
+                    "Verify the saved models."));
```

Comment:
Maybe, instead, `"Verify that there is a model in the datastore that matches the provided predicate."` ?